### ISSUE TITLE
Server icon/label customization

### ISSUE THREAD

**Issue by MitchTalmadge** [brandly/Lax#99]
Just a fun idea. Right now the label is a number:

![image](https://user-images.githubusercontent.com/8848080/92440916-80536d80-f16a-11ea-8b62-ddc22ddb67ce.png)

Perhaps it could be customized, like "fn" for freenode. Or maybe users could upload a picture somehow? Dunno. Curious to hear your thoughts.

**Comment by brandly** [comment]
i think we should do better than just a number, and it'd be great to be able to distinguish your connections.

the "fn" idea is cool. you can connect to any domain or IP, so we'd have a couple cases to deal with. i'm not sure what the "freenode" -> "fn" function looks like that supports arbitrary domains. if you wanna play with this and the result looks good, i think it'd be a nice improvement!

**Comment by MitchTalmadge** [comment]
> i'm not sure what the "freenode" -> "fn" function looks like that supports arbitrary domains

Yeah that's the problem I was thinking about too... hmm. I will sit on this for a bit :P

**Comment by nineninesevenfour** [comment]
Hello guys,

I hope I may give my two cents and share an idea:
I personally love what Mozilla did on Firefox's "about:newtab" page: When they do not have an icon for a website, they use a thing called "letter fallback", which looks like this:

<img width="482" alt="LetterFallbackExample" src="https://user-images.githubusercontent.com/75562299/129779493-da2ac51e-669c-4c0b-9b6b-2aafb7a9786a.png">

It is a very short algorithm which produces a more or less unique combination of letter and color.
How they did it?
* build a simple hash code from the URL
* with that hash code choose a color from a predefined array
* apply some fancy CSS with ::before and the first letter

Ingredients:

1) the hash function:
```js
/*
 * Takes the url as a string, runs it through a simple (non-secure) hash turning it into a random number
 * Apply that random number to the color array. The same url will always generate the same color.
 */
function generateColor() {
    let { title, colors } = this.props;
    if (!colors) {
        return "";
    }

    let colorArray = colors.split(",");

    const hashStr = str => {
        let hash = 0;
        for (let i = 0; i < str.length; i++) {
            let charCode = str.charCodeAt(i);
            hash += charCode;
        }
        return hash;
    };

    let hash = hashStr(title);
    let index = hash % colorArray.length;
    return colorArray[index];
}
```
source: https://hg.mozilla.org/mozilla-central/file/tip/browser/components/newtab/content-src/components/TopSites/TopSite.jsx

2. their choice of colors: `#0090ED,#FF4F5F,#2AC3A2,#FF7139,#A172FF,#FFA437,#FF2A8A`
It is a nice selection of pastel colors with tendency to warm colors
(I could not find the source for it, I just debugged it)

3. their choice of fonts: `font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Ubuntu", "Helvetica Neue", sans-serif` (might be platform-dependent, IDK)

4. apply the CSS with the ::before pseudo-selector, of which the most important parts are:
```CSS
    font-size: 64px;
    font-weight: 800;
    transform: rotate(-10deg);
    top: 6px;
    position: relative;
    color: #FFF;
```
the full list is:
```CSS
{
    content: attr(data-fallback);
    text-transform: uppercase;
    display: flex;
    align-items: center;
    justify-content: center;
    font-size: 64px;
    font-weight: 800;
    transform: rotate(-10deg);
    top: 6px;
    position: relative;
    color: #FFF;
}
```
(I think the flex-box-related stuff is need only if in a flex box)
source: https://hg.mozilla.org/mozilla-central/file/tip/browser/components/newtab/content-src/components/TopSites/_TopSites.scss

Of course it can be adapted a little (change colors, font) and you should remove common parts like "chat." to not use letter C for all servers.

What do you think?

**Comment by nineninesevenfour** [comment]
I tested the algorithm with the strings "chat.freenode.net" and "irc.libera.chat", which would give this result:
<img width="227" alt="LetterFallbackExampleChat" src="https://user-images.githubusercontent.com/75562299/129785939-49459348-08d4-4c33-83b5-672db1db347b.png">
And in *Lax* it would look like this:
![LetterFallbackExampleChat2](https://user-images.githubusercontent.com/75562299/129786044-ee0b632e-59d1-4815-b01d-cbbab09ecf9e.png)

**Comment by nineninesevenfour** [comment]
Just in case you prefer the original (more minimalistic) idea of two letters suggest the following algorithm which logically splits the string into *vowels* and *consonants*:
```js
const vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'];

function isVowel(char) {
    return vowels.includes(char);
}

function twoLetters(string) {
    if (!string) {
        return '??';
    }
    if (typeof string !== 'string') {
        return twoLetters(typeof string);
    }
    if (string.length < 2) {
        return string.repeat(2);
    }
    // 1. take the first letter
    const first = string[0];
    // 2. search for the next vowel
    let i;
    for (i = 1; i < string.length && !isVowel(string[i]); i++);
    // in case we reached the end
    if (i >= string.length) {
        return first + string[i - 1];
    }
    // 3. seatch for the next consonant
    for (; i < string.length && isVowel(string[i]); i++);
    if (i >= string.length) {
        // we reached the end
        return first + string[i - 1];
    } else {
        // take that consonant
        return first + string[i];
    }
}
```
Of course this works only for Ascii.
A small test gives the follwing result:
```
"freenode" ==> fn
"libera"   ==> lb
"mozilla"  ==> mz
"lax"      ==> lx
"eeeff"    ==> ef
"fffee"    ==> fe
"eeeee"    ==> ee
"fffff"    ==> ff
"f"        ==> ff
""         ==> ??
*falsy*    ==> ??
42         ==> nm
```
For the letter *y* it's a bit a matter of taste I'd say - it's sometimes a vowel and sometimes a consonant: https://www.merriam-webster.com/words-at-play/why-y-is-sometimes-a-vowel-usage
If you treat the *y* as consonant:
```
"oyster"   ==> or
"yard"     ==> yr
"lawyer"   ==> lw
"beyond"   ==> by
```
If you treat the *y* as vowel:
```
"oyster"   ==> os
"yard"     ==> yr
"lawyer"   ==> lw
"beyond"   ==> bn
```
A correct distinction according to language rules would be a bit overkill here, I'd say 😆 .
